packages/regl-worldview/src/commands/Command.js

```diff
@@ -9,79 +9,75 @@
 import WorldviewReactContext from "../WorldviewReactContext";
 
 export type Props<T> = {
   children?: T[],
+  reglCommand: RawCommand<T>,
   getHitmapId?: (T) => ?number,
   layerIndex?: number,
 };
 
-type CommandProps = {
-  layerIndex?: number,
-};
-
-// The base class for all regl-based drawing commands.
-// Subclasses should override getDrawProps and (and optionally getHitmapProps).
-export default class Command<T: CommandProps> extends React.Component<T> {
+// Component to dispatch draw props and hitmap props and a reglCommand to the render loop to render with regl.
+export default class Command extends React.Component<Props<T>> {
   context: WorldviewContextType;
   static displayName = "Command";
-  static command: RawCommand<any>;
 
-  // do not override this in sub-commands
-  componentDidMount() {
-    this.context.onMount(this);
-    this._updateContext();
+  constructor(props) {
+    super(props);
+    // In development put a check in to make sure the reglCommand prop is not mutated.
+    // Similar to how react checks for unsupported or deprecated calls in a development build.
+    if (process && process.env && process.env.NODE_ENV !== "production") {
+      this.shouldComponentUpdate = (nextProps: Props) => {
+        if (nextProps.reglCommand !== this.props.reglCommand) {
+          console.error("Changing the regl command prop on a <Command /> is not supported.");
+        }
+        return true;
+      };
+    }
   }
 
-  // do not override this in sub-commands
-  componentDidUpdate() {
-    this._updateContext();
+  componentDidMount() {
+    this.context.onMount(this, this.props.reglCommand);
   }
 
-  // do not override this in sub-commands
   componentWillUnmount() {
     this.context.onUnmount(this);
   }
 
-  _updateContext() {
-    const drawProps = this.getDrawProps();
+  _updateContext(context: WorldviewContextType) {
+    const drawProps = this.props.drawProps;
     if (drawProps == null) {
       return;
     }
-    this.context.registerDrawCall({
-      command: this,
+    context.registerDrawCall({
+      instance: this,
+      command: this.props.reglCommand,
       drawProps,
       layerIndex: this.props.layerIndex,
     });
 
-    const hitmapProps = this.getHitmapProps();
+    const hitmapProps = this.props.hitmapProps;
     if (hitmapProps) {
-      this.context.registerHitmapCall({
-        command: this,
+      context.registerHitmapCall({
+        instance: this,
+        command: this.props.reglCommand,
         drawProps: hitmapProps,
         layerIndex: this.props.layerIndex,
       });
     }
   }
 
-  getDrawProps(): any {
-    throw new Error("Overriding getDrawProps in subclass is required");
-  }
-
-  getHitmapProps(): any {
-    return null;
-  }
-
   render() {
     return (
       <WorldviewReactContext.Consumer>
         {(ctx: ?WorldviewContextType) => {
           if (ctx) {
             this.context = ctx;
+            this._updateContext(ctx);
```

Comment:
Why move this here from `componentDidUpdate`?

Reply:
I moved it back